### PR TITLE
Smallfiles test

### PR DESCRIPTION
**Pull request by spuiuk**
Adding a smallfiles test which uses podman. 

This will not work at this moment without installing the podman packages and disabling selinux. So setting it to draft for now.

### CODE REVIEW

**Review by synarete — COMMENTED** [review]
Why merge 'main' into 'smallfiles' (and not rebase)?

**Comment by spuiuk** [comment]
> Why merge 'main' into 'smallfiles' (and not rebase)?

I think that is what you get when you rebase from the github interface.

**Comment by spuiuk** [comment]
Pushed the latest set of patches. the following changes were done.

1) Use pathlib.Path - Some of the testhelper functions called return a string or expect a string. For such cases, you will see conversions being made to Path() and to str(). I will post patches after this where we move the helper functions to using the pathlib.Path where these awkward conversions should no longer be needed.

2) Change the way test_info and container_tests global variables are set. The earlier way to set the global variables were a result of having to accommodate both pytest and running it as standalone. This is no longer needed after we moved completely to pytest.

3) Changed the smallfiles container to use stream9. Uploaded this to the quay.io repo.

File: testcases/containers/images/smallfile/Containerfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#Version: 0.0.1
+FROM	fedora:38
+ENV	REFRESHED_AT 2020-08-06-01
+
+RUN	dnf update -y
+RUN	dnf install -y git python
```
[the remark this answers is not in the document]
Comment (spuiuk) [reply testcases/containers/images/smallfile/Containerfile:6]: The tests we run don't seem to need them and are executed without any problems. You can see the output of the smallfiles test in the resulting console output.

File: testcases/containers/images/smallfile/Containerfile [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#Version: 0.0.1
+FROM	fedora:38
```
[the remark this answers is not in the document]
Comment (spuiuk) [reply testcases/containers/images/smallfile/Containerfile]: done.

File: testcases/containers/images/smallfile/Makefile [hunk 3]
```diff
@@ -0,0 +1,8 @@
+CONTAINER_NAME := sit-test-cases:smallfiles
+build:
+	podman build . -t $(CONTAINER_NAME)
+push:
```
[the remark this answers is not in the document]
Comment (spuiuk) [reply testcases/containers/images/smallfile/Makefile:6]: done

File: testcases/containers/test_containers.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+import shutil
+import yaml
+
+script_root = os.path.dirname(os.path.realpath(__file__))
+container_tests_file = os.path.join(script_root, "test_containers.yml")
+
+# Use a global test_info to get a better output when running pytest
+test_info: typing.Dict[str, typing.Any] = {}
+# global containing tests read from yaml
+container_tests: typing.Dict[str, str] = {}
```
[the remark this answers is not in the document]
Comment (spuiuk) [reply testcases/containers/test_containers.py:22]: Both test_info and container_tests are put in as global. I have however changed the way these are populated. 

Both globals are instantiated only once but used more than once in different functions. For the sake of simplicity, we can leave them as globals.

File: testcases/containers/test_containers.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+import testhelper
+import os
+import pytest
+import typing
+import shutil
+import yaml
+
+script_root = os.path.dirname(os.path.realpath(__file__))
+container_tests_file = os.path.join(script_root, "test_containers.yml")
```
[the remark this answers is not in the document]
Comment (spuiuk) [reply testcases/containers/test_containers.py]: done

File: testcases/containers/test_containers.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+# ip addresses).
+
+import testhelper
+import os
+import pytest
+import typing
+import shutil
+import yaml
+
+script_root = os.path.dirname(os.path.realpath(__file__))
```
[the remark this answers is not in the document]
Comment (spuiuk) [reply testcases/containers/test_containers.py]: done

File: testcases/containers/test_containers.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+container_tests_file = os.path.join(script_root, "test_containers.yml")
+
+# Use a global test_info to get a better output when running pytest
+test_info: typing.Dict[str, typing.Any] = {}
+# global containing tests read from yaml
+container_tests: typing.Dict[str, str] = {}
+
+
+def containers_check(
+    mount_point: str, ipaddr: str, share_name: str, test: str
```
[the remark this answers is not in the document]
Comment (spuiuk) [reply testcases/containers/test_containers.py]: done

File: testhelper/cmdhelper.py [hunk 8]
```diff
@@ -81,3 +81,48 @@ def smbclient(
...
+
+    Parameters:
+    test_image: The image to be used for the podman run
+    test_root: The root of the folder which will be used to perform the tests
+
+    Returns:
+    int: Return value from the execution
+    string: stdout
+    """
+    cmds = ["/usr/bin/podman", "/usr/bin/docker"]
```
[the remark this answers is not in the document]
Comment (spuiuk) [reply testhelper/cmdhelper.py]: done

File: testhelper/cmdhelper.py [hunk 9]
```diff
@@ -81,3 +81,48 @@ def smbclient(
...
+    Returns:
+    str: first available file path from the list
+    """
+    for c in cmds:
+        if os.path.exists(c):
+            return c
+    return ""
+
+
+def podman_run(test_image: str, test_root: str) -> typing.Tuple[int, str]:
```
[the remark this answers is not in the document]
Comment (spuiuk) [reply testhelper/cmdhelper.py]: done

File: testhelper/cmdhelper.py [hunk 10]
```diff
@@ -81,3 +81,48 @@ def smbclient(
...
+def check_cmds(cmds: typing.List[str]) -> str:
+    """Return first file path which exists.
+
+    Parameters:
+    cmds: list of commands to test
+
+    Returns:
+    str: first available file path from the list
+    """
+    for c in cmds:
```
[the remark this answers is not in the document]
Comment (spuiuk) [reply testhelper/cmdhelper.py:97]: done

File: testhelper/cmdhelper.py [hunk 11]
```diff
@@ -81,3 +81,48 @@ def smbclient(
         stderr=subprocess.STDOUT,
     )
     return (ret.returncode, ret.stdout)
+
+
+def check_cmds(cmds: typing.List[str]) -> str:
```
[the remark this answers is not in the document]
Comment (spuiuk) [reply testhelper/cmdhelper.py]: done

**Review by synarete — APPROVED** [review]

**Comment by spuiuk** [comment]
Output of the smallfiles test from the centos-ci/xfs tests

```
       "==================================== PASSES ====================================",
        "_________ test_containers[192.168.123.10-share-xfs-default-smallfiles] _________",
        "----------------------------- Captured stdout call -----------------------------",
        "Trying to pull quay.io/samba.org/sit-test-cases:smallfiles...",
        "Getting image source signatures",
        "Copying blob sha256:b7954271d65c4abe8e117cb3f1497cf02139d069933b1223e5b5ad01e094458a",
        "Copying blob sha256:7a80604fbc801b1ca549efebfce436cd52028affa9e073b47f6260dfa853bf57",
        "Copying blob sha256:35f8654d391b38bcded9291f440d06fc3b0c5dccc0d8361c36e2c0bb32b00fc5",
        "Copying blob sha256:1926f8a6af8b3e39edf1650fc135e57b5745658f6b4977b1cf880f732bd29bcc",
        "Copying blob sha256:5218344d3a8991890ee4ad90c322698913a4d059db6bad4fc8cb1bb0c9c3b6c5",
        "Copying blob sha256:04e393290efc212a09612383affe45dadb2d3383f2229a68bf24e563a169bbe0",
        "Copying blob sha256:6476010ca3f151058a7c396ae04bf498d7b9aedda02a23cf20a8fb32a5dacb3a",
        "Copying config sha256:e7b01d2ce37f68d6da26d83dc636b47308c9db1fc9c486c19ebfb1d31409ef32",
        "Writing manifest to image destination",
        "changing --finish to true for op type create",
        "                                 version : 3.2",
        "                           hosts in test : None",
        "                        launch by daemon : False",
        "                   top test directory(s) : ['/testdir']",
        "                               operation : create",
        "                            files/thread : 1024",
        "                                 threads : 4",
        "           record size (KB, 0 = maximum) : 0",
        "                          file size (KB) : 1024",
        "                  file size distribution : fixed",
        "                           files per dir : 100",
        "                            dirs per dir : 10",
        "              threads share directories? : N",
        "                         filename prefix : ",
        "                         filename suffix : ",
        "             hash file number into dir.? : N",
        "                     fsync after modify? : Y",
        "                         incompressible? : N",
        "          pause between files (microsec) : 0",
        "                             auto-pause? : N",
        " delay after cleanup per file (microsec) : 0",
        "             minimum directories per sec : 2",
        "                             total hosts : 1",
        "                    finish all requests? : Y",
        "                              stonewall? : Y",
        "                 measure response times? : N",
        "                            verify read? : Y",
        "                                verbose? : N",
        "                          log to stderr? : N",
        "host = 5db1b9173a5b,thr = 00,elapsed = 44.94384574890137,files = 1024,records = 1024,status = ok",
        "host = 5db1b9173a5b,thr = 01,elapsed = 44.910395860672,files = 1024,records = 1024,status = ok",
        "host = 5db1b9173a5b,thr = 02,elapsed = 44.877068281173706,files = 1024,records = 1024,status = ok",
        "host = 5db1b9173a5b,thr = 03,elapsed = 44.87704110145569,files = 1024,records = 1024,status = ok",
        "total threads = 4",
        "total files = 4096",
        "total IOPS = 91",
        "total data =     4.000 GiB",
        "100.00% of requested files processed, warning threshold is  70.00%",
        "elapsed time =    44.944",
        "files/sec = 91.220737",
        "IOPS = 91.220737",
        "MiB/sec = 91.220737",
        "                                 version : 3.2",
        "                           hosts in test : None",
        "                        launch by daemon : False",
        "                   top test directory(s) : ['/testdir']",
        "                               operation : delete",
        "                            files/thread : 1024",
        "                                 threads : 4",
        "           record size (KB, 0 = maximum) : 0",
        "                          file size (KB) : 1024",
        "                  file size distribution : fixed",
        "                           files per dir : 100",
        "                            dirs per dir : 10",
        "              threads share directories? : N",
        "                         filename prefix : ",
        "                         filename suffix : ",
        "             hash file number into dir.? : N",
        "                     fsync after modify? : Y",
        "                         incompressible? : N",
        "          pause between files (microsec) : 0",
        "                             auto-pause? : N",
        " delay after cleanup per file (microsec) : 0",
        "             minimum directories per sec : 2",
        "                             total hosts : 1",
        "                    finish all requests? : N",
        "                              stonewall? : Y",
        "                 measure response times? : N",
        "                            verify read? : Y",
        "                                verbose? : N",
        "                          log to stderr? : N",
        "host = 5db1b9173a5b,thr = 00,elapsed = 3.500145435333252,files = 1024,records = 0,status = ok",
        "host = 5db1b9173a5b,thr = 01,elapsed = 3.5007212162017822,files = 1024,records = 0,status = ok",
        "host = 5db1b9173a5b,thr = 02,elapsed = 3.4994964599609375,files = 1024,records = 0,status = ok",
        "host = 5db1b9173a5b,thr = 03,elapsed = 3.5004777908325195,files = 1024,records = 0,status = ok",
        "total threads = 4",
        "total files = 4096",
        "100.00% of requested files processed, warning threshold is  70.00%",
        "elapsed time =     3.501",
        "files/sec = 1170.215446",
```

**Comment by spuiuk** [comment]
merging based on output from the centos-ci/xfs test only.
(https://github.com/samba-in-kubernetes/sit-environment/issues/53)

**Pull request merged**